spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/actions/GenerateSymlinkManifestAction.java

```diff
@@ -0,0 +1,145 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.actions;
+
+import java.util.Arrays;
+import java.util.List;
+import org.apache.iceberg.MetadataTableType;
+import org.apache.iceberg.Partitioning;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.actions.GenerateSymlinkManifest;
+import org.apache.iceberg.actions.GenerateSymlinkManifestActionResult;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.spark.JobGroupInfo;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.DataFrameWriter;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+
+public class GenerateSymlinkManifestAction extends BaseSparkAction<GenerateSymlinkManifestAction>
+    implements GenerateSymlinkManifest {
+
+  private final Table table;
+  private boolean ignoreDeleteFiles = false;
+  private String rootLocation;
+
+  private static final String SYMLINK_MANIFEST_DEFAULT_DIRECTORY = "_symlink_format_manifest";
+
+  protected GenerateSymlinkManifestAction(SparkSession spark, Table table) {
+    super(spark);
+    ValidationException.check(
+        table.currentSnapshot() != null, "Cannot generate symlink manifest for empty table");
+    this.table = table;
+  }
+
+  @Override
+  public GenerateSymlinkManifestAction.Result execute() {
+    JobGroupInfo info = newJobGroupInfo("GENERATE-SYMLINK-MANIFEST", jobDesc());
+    return withJobGroupInfo(info, this::doExecute);
+  }
+
+  @Override
+  protected GenerateSymlinkManifestAction self() {
+    return this;
+  }
+
+  @Override
+  public GenerateSymlinkManifest ignoreDeleteFiles() {
+    this.ignoreDeleteFiles = true;
+    return this;
+  }
+
+  @Override
+  public GenerateSymlinkManifest symlinkManifestRootLocation(String rootLocation) {
+    this.rootLocation = rootLocation;
+    return this;
+  }
+
+  private GenerateSymlinkManifest.Result doExecute() {
+    long snapshot = table.currentSnapshot().snapshotId();
+    Dataset<Row> entries =
+        SparkTableUtil.loadMetadataTable(spark(), table, MetadataTableType.ENTRIES);
+    // If we're not ignoring delete files, check if there are any delete manifests for the snapshot.
+    if (!ignoreDeleteFiles) {
+      boolean deletesExist = entries.filter("status < 2 and data_file.content > 0").isEmpty();
+      if (deletesExist) {
+        throw new UnsupportedOperationException(
+            "Cannot generate symlink manifest when there are delete files. Set ignore delete files for generating a "
+                + "symlink manifest of only data files");
```

Comment:
@szehon-ho @aokolnychyi @jackye1995 Appreciate any inputs on if there is a more efficient way to just validate if there are delete files beyond filtering on the whole entries table. basically, I'm still checking is there a way at higher layers of the metadata tree to filter for the existence of delete files for a given table snapshot rather than having to do a filter on the list of data/delete files in the current table state?

My thinking is to leverage the AllManifests metadata table and filter based on reference_snapshot_id and the manifest type. Since a manifest file can only contain deletes or only contain data files, I filtered if there are any delete manifests for the given reference_snapshot_Id but reference_snapshot_Id is the snapshot when that manifest is produced which isn't what we really want.

Reply:
Are we actually able to query symlink table with deletes? My understanding is that symlink table will produce something like a Hudi read optimized view that does not have the deletes applied at all. If this is a pre-condition, we don't really need to add this check.